### PR TITLE
Restock supports custom hometowns for buying items.

### PR DESCRIPTION
**Pull request by jayRyan24**
Initial commit for adding custom hometowns to restock. If a hometown is found, it will then go to that hometown instead. I added two new money methods to handle things. Also fixed a bug in the base consumables for hib's items needing price and not cost.

### CODE REVIEW

**Comment by Sarvatt** [comment]
the common-money change should be broken out into it's own commit that goes in a few days before the rest to not break people who log in new characters with an existing one still logged in

**Comment by jayRyan24** [comment]
That's a lot of work. If they break, they just need to announce it and then start over again.

**Comment by jayRyan24** [comment]
I removed the common-money methods which are called and put them in a separate PR. They are located in #2801 .

File: restock.lic [hunk 1]
```diff
@@ -1,58 +1,65 @@
...
       num_needed = item['quantity'] - remaining
       buy_num = (num_needed / item['size'].to_f).ceil
       coin_needed += buy_num * item['price']
       item['buy_num'] = buy_num
       items_to_restock.push(item)
     end
-
-    ensure_copper_on_hand(coin_needed, @settings) if coin_needed > 0
-
+    town = @settings.hometown
```
Comment (rpherbig) [inline restock.lic]: ❓ What would you think about having `town` be a second parameter to `restock_items`? Then line 14 would pass in `@settings.hometown` and line 20 would pass in `hometown`

File: restock.lic [hunk 2]
```diff
@@ -1,58 +1,65 @@
...
     setup
-
     items = parse_restockable_items
+
+    custom_loc_items = items.select{ |k| k['hometown'] }
+
+    items -= custom_loc_items
+    restock_items(items)
+
+    unique_hometowns = custom_loc_items.map{ |k| "#{k['hometown']}" }.uniq
```
Comment (rpherbig) [inline restock.lic]: ❓ Why is there string interpolation here instead of mapping to `k['hometown']` directly?

File: restock.lic [hunk 3]
```diff
@@ -1,58 +1,65 @@
...
+    unique_hometowns.each do |hometown|
```
Comment (rpherbig) [inline restock.lic]: 💡 You don't need `unique_hometowns` as a variable. Method chaining here could look like
```ruby
custom_loc_items.map{ |k| "#{k['hometown']}" }
.uniq
.each do ...
```

File: restock.lic (hunk 2 above)
Comment (jayRyan24) [reply restock.lic]: Not sure, but I fixed it.

File: restock.lic (hunk 3 above)
Comment (jayRyan24) [reply restock.lic]: Yup, I used it differently before this, so I had a use of it then. Thanks!

File: restock.lic [hunk 4]
```diff
@@ -1,58 +1,65 @@
...
+      custom_list = custom_loc_items.select{ |k| k['hometown'].eql?(hometown) }
```
[the remark this answers is not in the document]
Comment (jayRyan24) [reply restock.lic]: ok, so it didn't do it in that order this time, so yeah, doesn't even matter

File: restock.lic (hunk 1 above)
Comment (jayRyan24) [reply restock.lic]: Yup, good change. Thank you.

File: restock.lic [hunk 5]
```diff
@@ -85,25 +90,21 @@ class Restock
...
   end
-
   def valid_item_data?(item_data)
-    return false unless item_data.key?('name')
-    return false unless item_data.key?('size')
-    return false unless item_data.key?('room')
-    return false unless item_data.key?('price')
-    return false unless item_data.key?('stackable')
-    return false unless item_data.key?('quantity')
+    return false unless item_data.key?('name')     &&\
```
[the remark this answers is not in the document]
Comment (jayRyan24) [reply restock.lic]: Sure thing.

**Comment by jayRyan24** [comment]
Yup, that's the plan Stan. Thanks for the review!

**Comment by jayRyan24** [comment]
Oh, tested all the changes. Continues to work.

File: restock.lic [hunk 6]
```diff
@@ -37,10 +36,8 @@ class Restock
       item['buy_num'] = buy_num
       items_to_restock.push(item)
     end
-    town = @settings.hometown
-    town = items_to_restock.first['hometown'] if items_to_restock.first.key?('hometown')
-    coin_string = coin_needed.to_s + ' copper ' + town_currency(town)
-    get_money_from_specific_bank(coin_string, town) if coin_needed > 0
+    money_string = coin_needed.to_s + ' copper ' + town_currency(town)
```
Comment (rpherbig) [inline restock.lic]: 💡 This is a case where I prefer string interpolation, something like `"#{coin_needed.to_s} copper #{town_currency(town)}"`

**Comment by rpherbig** [comment]
Looks good to me. Remind me to merge this at some point 😁

File: restock.lic (hunk 6 above)
Comment (jayRyan24) [reply restock.lic]: Just so used to stringstream, lol. I gotcha.

**Comment by jayRyan24** [comment]
I've added this to https://elanthipedia.play.net/mediawiki/index.php/Lich_script_repository

**Comment by jayRyan24** [comment]
@rpherbig this can be pushed any time, it's been a few days.

**Pull request merged**